ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ImmDB.hs

```diff
@@ -496,31 +495,34 @@ stream db registry blockComponent from to = runExceptT $ do
           ImmDB.WrongHashError slot _ _ -> slot
 
     -- | The ImmutableDB doesn't support an exclusive end bound, so we stop
-    -- the iterator when it reaches its exclusive end bound.
-    stopAt :: forall a.
-              StreamTo blk
-           -> ImmDB.Iterator (HeaderHash blk) m (HeaderHash blk, a)
-           -> ImmDB.Iterator (HeaderHash blk) m (HeaderHash blk, a)
+    -- the iterator when the next block is the exclusive end bound.
+    stopAt :: StreamTo blk
+           -> ImmDB.Iterator (HeaderHash blk) m b
+           -> m (ImmDB.Iterator (HeaderHash blk) m b)
     stopAt = \case
-      StreamToInclusive _  -> id
-      StreamToExclusive pt -> \it -> it
-          { ImmDB.iteratorNext    = ignoreExclusiveBound <$> ImmDB.iteratorNext it
-          , ImmDB.iteratorPeek    = ignoreExclusiveBound <$> ImmDB.iteratorPeek it
-          , ImmDB.iteratorHasNext = ImmDB.iteratorHasNext it >>= \case
-              Nothing -> return Nothing
-              Just next@(_epochOrSlot, hash)
-                | isEnd hash
-                -> ImmDB.iteratorClose it $> Nothing
-                | otherwise
-                -> return $ Just next
-          }
+      StreamToInclusive _     -> return
+      StreamToExclusive endPt -> \it -> do
+          -- Whenever the iterator is moved to the next block, we check
+          -- whether the next block (cheap with 'ImmDB.iteratorHasNext') is
+          -- the end bound, in which case we close the iterator. All other
+          -- operations on the iterator can remain the same.
+          --
+          -- We must do the check immediately, because the very first block
+          -- might be the exclusive bound.
+          closeWhenEndIsNext it
+          return it
```

Comment:
Problem: when moving to the block right before the exclusive bound, we close the iterator, which means that `iteratorHasPeek` will return `IteratorExhausted` while it should return the current block.

Plan: remove `iteratorPeek` because it is no longer used!